### ISSUE TITLE
[doc] 5.2.5  Cluster Bootstrap

### ISSUE THREAD

**Issue by gghh** [SUSE/doc-caasp#739]
5.2.5  Cluster Bootstrap

https://documentation.suse.com/suse-caasp/4.0/html/caasp-deployment/bootstrap.html#cluster.bootstrap

> 3. Add additional master nodes to the cluster.
> Replace the `<IP/FQDN>` with the IP for the machine. Replace `<NODE_NAME>` with a unique identifier, for example, "master-two".
> ```
> skuba node join --role master --user sles --sudo --target <IP/FQDN> <NODE_NAME>
> ```

The documentation fails to mention that `<NODE_NAME>` needs a DNS entry so that other nodes can resolve it. In other words, from any node of the cluster, the command
```
nslookup NODE_NAME
```
should return the IP address of the node. This is also omitted in the man page for `skuba-node-bootstrap(1)` and `skuba-node-join(1)`, to which this remark applies as well

**Comment by gghh** [comment]
For the records, the fact that `skuba-node-bootstrap(1)` and `skuba-node-join(1)` require both a Fully Qualified Domain Name (FQDN) and a resolvable name `NODE_NAME` is confusing; a novice to the tool such as myself would assume that the tool can infer `NODE_NAME` from `FQDN` by itself. All tutorials I'm aware of indeed use nodes hostnames as `NODE_NAME`, such as
```
skuba node join --role master --user sles --sudo --target fred.example.com fred
```
Note the redundancy above, where one need to specify both `fred.example.com` and `fred`. If one tried to exploit the additional degree of freedom offered by the command syntax such as
```
skuba node join --role master --user sles --sudo --target fred.example.com paul
```
without the appropriate DNS configuration that would result in a broken deployment in a way that is not entirely obvious.

This is to say: a justification as to why it is possible to insert "a unique identifier, for example, master-two" (quoting the docs) would contribute to clarity.

**Comment by ereslibre** [comment]
The `--target` is the way the node from where you are bootstrapping from can SSH into the node, whereas the node name is the name that will node will register with against Kubernetes. They have different purposes, e.g. you could use `--target` with an IP address, while still give the node name a meaningful name.

What do you mean that it would result in a broken deployment if the DNS configuration is not in place? Can you please elaborate?

**Comment by r0ckarong** [comment]
> The documentation fails to mention that `<NODE_NAME>` needs a DNS entry so that other nodes can resolve it.

Because that isn't the case. `NODE_NAME` is simply a human readable moniker you can give to nodes so the outputs of `kubectl` or `skuba` make more sense to a person. The cluster will rely on the FQDN/IP of the nodes to communicate. Nothing will resolve, or try to resolve, `NODE_NAME` since it's not used in this way. 

You are free to create DNS entries pointing to your nodes that have the same string as what you used for the `NODE_NAME` but that's not necessary and I am not aware that this is an intended practice.

@ereslibre Please correct me if I'm wrong.

**Comment by r0ckarong** [comment]
> The `--target` is the way the node from where you are bootstrapping from can SSH into the target node, whereas the node name is the name that the target node will register with against Kubernetes. They have different purposes, e.g. you could use `--target` with an IP address, while still give the node name a meaningful name.
> 
> What do you mean that it would result in a broken deployment if the DNS configuration is not in place? Can you please elaborate?

I think the misunderstanding is that `NODE_NAME` becomes a "shorthand" for a node so using skuba/kubectl on it is easier. So instead of using `skuba update mylongname.mydomain.org` you would use `skuba update shorthand`. But that isn't implemented as far as I know.

**Comment by ereslibre** [comment]
I would like to know what exactly breaks if you use a nonresolvable DNS name for a given node as reported by @gghh. `kubeadm` by default sets:

`--kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname`

on the APIServer arguments, so when you will get proxied to kubelets, the `hostname` is unlikely to be used. For all we know at this time, the nodename can be descriptive, and I cannot envision any further issue if it's not resolvable by other nodes.

Internally, `kubeadm` will use discovered IP addresses to add `etcd` members to the cluster, so I cannot really foresee what could be broken under this assumption.

**Comment by gghh** [comment]
> > The documentation fails to mention that <NODE_NAME> needs a DNS entry so that other nodes can resolve it.
>
> Because that isn't the case. NODE_NAME is simply a human readable moniker you can give to nodes so the outputs of kubectl or skuba make more sense to a person. The cluster will rely on the FQDN/IP of the nodes to communicate. Nothing will resolve, or try to resolve, NODE_NAME since it's not used in this way.

Ok, thank you for clarifying this. The issue I opened is most likely invalid.

The reason I opened this issue is that it's been a while I'm trying to deploy my first cluster ever, `skube node join` is failing, a large number of warning and errors end up in my system logs. Some of the most prominent are along the lines of

```
[WARNING Hostname]: hostname "master0" could not be reached
[WARNING Hostname]: hostname "master0": lookup master0 on 10.100.2.10:53: no such host
```

and people around me started suggesting I look at DNS. Now, I've probably misinterpreted their suggestion understanding that the mnemonic name needs to DNS-resolve. Looking back at various tutorial I've found invocations such as

```
skuba node join --role master --target foo-worker-03.example.com foo-worker-03
```

(note: `NODE_NAME` in the example above would DNS-resolve if you have `example.com` in your domain search list). At this point I looked at all my invocations, which are like

```
skuba node join --role master --target foo-worker-03.example.com worker3
```

(note: `NODE_NAME` doesn't resolve) and I thought: There! This is my mistake! going back to the docs and blaming it.

After I opened this issue I tried with resolvable `NODE_NAME`s and surprise surprise my problem was NOT solved, still "nodeXY could not be reached" messages in the logs.

Thanks for the clarification, feel free to close. I'll go back to my broken cluster, at least I can cross this one out from my troubleshooting list.

**Comment by r0ckarong** [comment]
The `NODE_NAME` seems to be misleading a lot of new users. Reopening to find a better wording.

![Screenshot_20200320_174433](https://user-images.githubusercontent.com/4920542/77186055-92fdbb80-6ad2-11ea-8d69-74b73cc3bcaa.png)

**Comment by r0ckarong** [comment]
Find a better place to put the notification about what to set the node names to and make clear that while they are originally just a "random" name you can give to nodes, in Openstack and AWS you MUST use the same names that you gave the instances on which the nodes are running. Otherwise the CPI will try to resolve the names given in skuba and won't be able to match/find them in the instances.

**Comment by r0ckarong** [comment]
@flavio @ereslibre There seems to be a "hidden" requirement by skuba. Skuba will try and resolve the `NODE_NAME` that we teach users to create at will to an actually existing "node name" in the cloud provider instance pool.

- Is this intentional? 
- Is there maybe a better way to handle this in skuba? 
- Should we just update the docs saying "You MUST name the nodes the same thing that you are going to use for "NODE_NAME" in all of your Kubernetes outputs later"?

**Comment by ereslibre** [comment]
The nodename can be literally anything in general. Of course, it's a better approach if it's resolvable. Now, we have to define resolvable from where. Let me elaborate:

* API Server -> Kubelet tunneling: the API server needs to be able to resolve the name of the nodename, in order to establish a tunnel (e.g. when you do a `kubectl exec`).

  * By default, the API Server `kubelet-preferred-address-types` is: `InternalIP,ExternalIP,Hostname`. This can be changed by our users though, but in general, this means that the API Server will try to first use the Internal IP of the kubelet, then the External IP, then whatever the DNS server resolves the Hostname to in order to create this tunnel.

* CPI integration: the cloud provider components loops will add some metadata to nodes, information that is coming from the cloud provider itself. In order to be able to match nodes to VM's running in that cloud, it needs to use something in order to query, this is the nodename in this case. If a nodename does not match the name of the VM in that cloud provider, it won't be found, and thus, the reconciliation for that node will fail; so in this case, it is **required** that your nodenames match the name of the nodes on that cloud provider. There was a [proposed mitigation on the RFC](https://github.com/SUSE/caasp-rfc/blob/master/accepted/0024-openstack-cloud-provider-integration.md#node-names).

**Comment by r0ckarong** [comment]
The information about the node metadata was actually introduced with #541 just recently and shipped with 4.1.2. I will close this issue.
